release-21.2: release-22.1: backupccl: elide spans from backups that were subsequently reintroduced

msbutler opened the pull request:
Backport 2/2 commits from #88488.

/cc @cockroachdb/release

---

Backport 1/1 commits from #87312.

/cc @cockroachdb/release

---

Currently RESTORE may restore invalid backup data from a backed up table that
underwent an IMPORT rollback. See https://github.com/cockroachdb/cockroach/issues/87305 for a detailed explanation.

This patch ensures that RESTORE elides older backup data that were deleted via
a non-MVCC operation. Because incremental backups always reintroduce spans
(i.e. backs them up from timestamp 0) that may have undergone a non-mvcc
operation, restore can identify restoring spans with potentially corrupt data
in the backup chain and only ingest the spans' reintroduced data to any system
time, without the corrupt data.

Here's the basic impliemenation in Restore:
- For each span we want to restore
   - identify the last time, l, the span was introduced, using the manifests
   - dont restore the span using a backup if backup.EndTime < l

This implementation rests on the following assumption: the input spans for each
restoration flow (created in createImportingDescriptors) and the
restoreSpanEntries (created by makeSimpleImportSpans) do not span across
multiple tables. Given this assumption, makeSimpleImportSpans skips adding
files from a backups for a given input span that was reintroduced in a
subsequent backup.

It's worth noting that all significant refactoring occurs on code run by
the restore coordinator; therefore, no special care needs to be taken for
mixed / cross version backups. In other words, if the coordinator has updated,
the cluster restores properly; else, the bug will exist on the restored cluster.
It's also worth noting that other forms of this bug are apparent on older
cluster versions (https://github.com/cockroachdb/cockroach/issues/88042, https://github.com/cockroachdb/cockroach/issues/88043) and has not been noticed by customers; thus,
there is no need to fail a mixed version restore to protect the customer from
this already existing bug.

Informs https://github.com/cockroachdb/cockroach/issues/87305

Release justification: bug fix

Release note: fix for TA advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

msbutler committed:
backupccl: elide spans from backups that were subsequently reintroduced

Currently RESTORE may restore invalid backup data from a backed up table that
underwent an IMPORT rollback. See #87305 for a detailed explanation.

This patch ensures that RESTORE elides older backup data that were deleted via
a non-MVCC operation. Because incremental backups always reintroduce spans
(i.e. backs them up from timestamp 0) that may have undergone a non-mvcc
operation, restore can identify restoring spans with potentially corrupt data
in the backup chain and only ingest the spans' reintroduced data to any system
time, without the corrupt data.

Here's the basic impliemenation in Restore:
- For each span we want to restore
   - identify the last time, l, the span was introduced, using the manifests
   - dont restore the span using a backup if backup.EndTime < l

This implementation rests on the following assumption: the input spans for each
restoration flow (created in createImportingDescriptors) and the
restoreSpanEntries (created by makeSimpleImportSpans) do not span across
multiple tables. Given this assumption, makeSimpleImportSpans skips adding
files from a backups for a given input span that was reintroduced in a
subsequent backup.

It's worth noting that all significant refactoring occurs on code run by
the restore coordinator; therefore, no special care needs to be taken for
mixed / cross version backups. In other words, if the coordinator has updated,
the cluster restores properly; else, the bug will exist on the restored cluster.
It's also worth noting that other forms of this bug are apparent on older
cluster versions (#88042, #88043) and has not been noticed by customers; thus,
there is no need to fail a mixed version restore to protect the customer from
this already existing bug.

Informs #87305

Release justification: bug fix

Release note (bug fix): fix for TA advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

msbutler committed:
backupccl: reintroduce previously offline tables with manifest.DescriptorChanges

getReintroducedSpans finds all tables included in the current and previous
backup that may have undergone a non-mvcc operation. The current backup will
then back up these tables' spans from ts = 0, as the previous backup may have
missed certain non-mvcc written.

To find these tables, getReintroducedSpans must find all tables covered in the
previous backup that were in the offline state at previous backup start time.

This function assumed that the previous backup's
manifest.Descriptors field would contain all tables covered in the previous
backup; however, while investigating #88042, we discovered that this assumption
is not correct. During revision history backups, a table with an in-progress
import (e.g. offline at backup time) can get backed up and included in
manifest.DescriptorChanges but not in manifest.Descriptors. This implies that
getReintroducedSpans missed reintroducing spans from this table, implying that
backup chains have missed backing up some data.

This patch fixes getReintroducedSpans to ensure it reintroduces tables included
in manifest.DescriptorChanges whose last revision brought the table offline. In
addition, this patch adds a check to restore planning which causes the restore
to fail immedidiately if any restore target is missing an introduced span.

Release note(bug fix): fix apart of Tech Advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

Release justification: bug fix

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89019)
<!-- Reviewable:end -->

msbutler commented:
The significant changes relative to the 22.1 patch are the use of aditya's fixtures to generate backups for the unit tests. So, the refactoring related to test infra and the two dd tests deserve a closer look.

I also plan to squash commits 3 and 4 into 2 and delete commit 5.

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data:
```diff
@@ -0,0 +1,237 @@
+3
+
+
+# Check that that backup chain with the intermediate incremental backup restores properly
```
nit: I think we need to say backup without revision history here?

msbutler replied:
ah. good catch!

adityamaru reviewed (dismissed):
Mostly LGTM, but I've only given the datadriven tests a close review. I'm assuming the rest backported relatively cleanly.

msbutler commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data:
```diff
@@ -0,0 +1,237 @@
+----
+
+
+# Check that the wildcard restore encounters the same errors
```
Nah. I think we're sufficiently covered.

msbutler commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-rollback:
```diff
@@ -0,0 +1,137 @@
+#   in-progress-missing-data)
+# - roll it back it back with non-mvcc clear range
+# - run an inc backup and ensure we reintroduce the table spans
+# (the fixtures are located in ccl/backupccl/testdata/backup-restore/restore_importing_tables_with_patch)
```
Done. dirs now roughly correspond to dd test file name.

msbutler committed:
backupccl: refine span introduction checker

This patch fixes two bugs in restore's span introduction checker:

Previously, the checker would return an error if an incremental backup
observed a new view. Because we back up a view's descriptor but not its span,
the restore checker previously expected the backup to contain an introduced
span for the new view, causing the checker to spuriously fail. This patch
filters views out of the checker.

Previously, the checker only checked for introduced spans for table descriptors
that were online at manfiest.Endtime. Now, in the spirit of
getReintroducedSpans(), the checker also checks
descriptors that may have been offline at manifest.Endtime, but may
have been online between manifest.StartTime and manifest.Endtime.

Release note (bug fix): prevents restore from spuriously failing

msbutler commented:
I've added a commit that filters views out of restore's checker, though perhaps it's better for this commit to get incorperated in https://github.com/cockroachdb/cockroach/pull/86713

msbutler commented:
i plan to squash commits 3 and 4 into commit 2.
